aws/credentials/stscreds/web_identity_provider.go

```diff
@@ -0,0 +1,100 @@
+	// a new error to be returned during Retrieve.
+	ErrCodeWebIdentityRetrievalErr = "WebIdentityRetrievalErr"
+)
+
+// now is used to return a time.Time object representing
+// the current time. This can be used to easily test and
+// compare test values.
+var now = func() time.Time {
+	return time.Now()
+}
+
+// WebIdentityRoleProvider is used to retrieve credentials using
+// an OIDC token.
+type WebIdentityRoleProvider struct {
+	credentials.Expiry
+
+	client       stsiface.STSAPI
+	ExpiryWindow time.Duration
+
+	tokenFilePath string
+	roleARN       string
+}
+
+// NewWebIdentityCredentials will return a new set of credentials with a given
+// configuration, role arn, and token file path.
+func NewWebIdentityCredentials(c client.ConfigProvider, roleARN, path string) *credentials.Credentials {
+	svc := sts.New(c)
+	p := NewWebIdentityRoleProvider(svc, roleARN, path)
+	return credentials.NewCredentials(p)
+}
+
+// NewWebIdentityRoleProvider will return a new WebIdentityRoleProvider with the
+// provided stsiface.STSAPI
+func NewWebIdentityRoleProvider(svc stsiface.STSAPI, roleARN, path string) *WebIdentityRoleProvider {
+	return &WebIdentityRoleProvider{
+		client:        svc,
+		tokenFilePath: path,
+		roleARN:       roleARN,
+	}
+}
+
+var emptyTokenFilePathErr = awserr.New(ErrCodeWebIdentityRetrievalErr, "'WebIdentityTokenFilePath' environment variable is empty", nil)
+var emptyRoleARNErr = awserr.New(ErrCodeWebIdentityRetrievalErr, "'WebIdentityRoleARN' environment variable is empty", nil)
+
+// Retrieve will attempt to assume a role from a token which is located at
+// 'WebIdentityTokenFilePath' specified destination and if that is empty an
+// error will be returned.
+func (p *WebIdentityRoleProvider) Retrieve() (credentials.Value, error) {
+	if len(p.tokenFilePath) == 0 {
+		return credentials.Value{}, emptyTokenFilePathErr
+	}
+
+	if len(p.roleARN) == 0 {
+		return credentials.Value{}, emptyRoleARNErr
+	}
+
+	b, err := ioutil.ReadFile(p.tokenFilePath)
+	if err != nil {
+		errMsg := fmt.Sprintf("unabled to read file at %s", p.tokenFilePath)
+		return credentials.Value{}, awserr.New(ErrCodeWebIdentityRetrievalErr, errMsg, err)
+	}
+
+	// session name is used to uniquely identify a session. This simply
+	// uses unix time in nanoseconds to uniquely identify sessions.
+	sessionName := strconv.FormatInt(now().UTC().UnixNano(), 10)
+	resp, err := p.client.AssumeRoleWithWebIdentity(&sts.AssumeRoleWithWebIdentityInput{
+		RoleArn:          &p.roleARN,
+		RoleSessionName:  &sessionName,
+		WebIdentityToken: aws.String(string(b)),
+	})
+	if err != nil {
+		return credentials.Value{}, awserr.New(ErrCodeWebIdentityRetrievalErr, "failed to retrieve credentials", err)
+	}
+
+	p.SetExpiration(aws.TimeValue(resp.Credentials.Expiration), p.ExpiryWindow)
+
+	value := credentials.Value{
+		AccessKeyID:     aws.StringValue(resp.Credentials.AccessKeyId),
+		SecretAccessKey: aws.StringValue(resp.Credentials.SecretAccessKey),
```

Comment:
Probably want to set the `SessionToken` value as well incase the `AssumeRoleWithWebIdentity` call returns it.

Reply:
As stated above, a SessionToken is a required field and should be added